react-dom 15.4.0 does not work in AMD enviroment

If I try to load react-dom with RequireJS, I get an error: "Mismatched anonymous define()".
Looking at the react-dom from the 15.4.0 distribution (https://cdnjs.cloudflare.com/ajax/libs/react/15.4.0/react-dom.js), I notice that there seem to be two UMD wrappers at the top.
This is probably related to the change in the packaging structure (#7164 and #7168).

Ah, I couldn't get it to work yesterday because the gist is broken. It should read `		      React.createElement('h1', {}, 'hello from react'),` instead of `React.createElement('h1', 'hello from react'),`
I can make it work by applying this to `react-dom.js`:
```diff
--- react-dom	2016-11-18 15:16:33.813492205 +0100
+++ react-dom.js	2016-11-18 15:18:09.505958843 +0100
@@ -9,7 +9,7 @@
 
   // RequireJS
   } else if (typeof define === "function" && define.amd) {
-    require(['react'], f);
+    define(['react'], f);
 
   // <script>
   } else {
@@ -29,7 +29,7 @@
     f(g.React)
   }
 })(function(React) {
-  (function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.ReactDOM = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(_dereq_,module,exports){
+  return (function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){return f()}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.ReactDOM = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(_dereq_,module,exports){
 /**
  * Copyright 2013-present, Facebook, Inc.
  * All rights reserved.
```

--- a/grunt/config/browserify.js
+++ b/grunt/config/browserify.js
@@ -74,7 +74,12 @@ function simpleBannerify(src) {
 // variable. So we write a wrapper around the UMD bundle that browserify creates,
 // and define a React variable that will require across Webpack-boundaries or fall
 // back to the global, just like it would previously.
+// Additionally, we're going to modify the original UMD wrapper a bit, replacing
+// the define([], f) call with f(). This allows our additional wrapper to work
+// so we can define() there without issue (AMD was causing issues). If you
+// think this is all a bit insane, you are absolutely correct.
 function wrapperify(src) {
+  src = src.replace('define([],f)', 'return f()');
   return `
 ;(function(f) {
   // CommonJS
@@ -83,7 +88,7 @@ function wrapperify(src) {
 
   // RequireJS
   } else if (typeof define === "function" && define.amd) {
-    require(['react'], f);
+    define(['react'], f);
 
   // <script>
   } else {
@@ -103,7 +108,7 @@ function wrapperify(src) {
     f(g.React)
   }
 })(function(React) {
-  ${src}
+  return ${src}
 });
 `;
 }

--- a/grunt/config/browserify.js
+++ b/grunt/config/browserify.js
@@ -63,7 +63,7 @@ function simpleBannerify(src) {
 }
 
 // What is happening here???
-// I'm glad you asked. It became really to make our bundle splitting work.
+// I'm glad you asked. It became really hard to make our bundle splitting work.
 // Everything is fine in node and when bundling with those packages, but when
 // using our pre-packaged files, the splitting didn't work. Specifically due to
 // the UMD wrappers defining their own require and creating their own encapsulated
@@ -71,20 +71,29 @@ function simpleBannerify(src) {
 // be smart and looks for top-level requires (even when aliasing to a bundle),
 // but since we didn't have those, we couldn't require 'react' from 'react-dom'.
 // But we are already shimming in some modules that look for a global React
-// variable. So we write a wrapper around the UMD bundle that browserify creates,
-// and define a React variable that will require across Webpack-boundaries or fall
-// back to the global, just like it would previously.
-// Additionally, we're going to modify the original UMD wrapper a bit, replacing
-// the define([], f) call with f(). This allows our additional wrapper to work
-// so we can define() there without issue (AMD was causing issues). If you
-// think this is all a bit insane, you are absolutely correct.
+// variable. So we replace the UMD wrapper that browserify creates with out own,
+// in 2 steps.
+// 1. We swap out the browserify UMD with a plain function call. This ensures
+// that the internal wrapper doesn't interact with the external state. By the
+// time we're in the internal wrapper it doesn't matter what the external wrapper
+// detected. Browserify insulates its CommonJS system inside closures so can just
+// call that function and return it.
+// 2. We put our own UMD wrapper around that fixed internal function, ensuring
+// React is in scope. This outer wrapper is essentially the same UMD wrapper
+// browserify would create, just handling the scope issue.
+// Is this insane? Yes.
+// Does it work? Yes.
+// Should it go away ASAP? Yes.
 function wrapperify(src) {
-  src = src.replace('define([],f)', 'return f()');
+  src = src.replace(
+    `function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.${this.data.standalone} = f()}}`,
+    `function(f){return f()}`
+  );
   return `
 ;(function(f) {
   // CommonJS
   if (typeof exports === "object" && typeof module !== "undefined") {
-    f(require('react'));
+    module.exports = f(require('react'));
 
   // RequireJS
   } else if (typeof define === "function" && define.amd) {
@@ -105,7 +114,7 @@ function wrapperify(src) {
       // see https://github.com/facebook/react/issues/3037
       g = this;
     }
-    f(g.React)
+    g.${this.data.standalone} = f(g.React);
   }
 })(function(React) {
   return ${src}

--- a/grunt/config/browserify.js
+++ b/grunt/config/browserify.js
@@ -85,8 +85,13 @@ function simpleBannerify(src) {
 // Does it work? Yes.
 // Should it go away ASAP? Yes.
 function wrapperify(src) {
+  var toReplace =
+    `function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.${this.data.standalone} = f()}}`;
+  if (src.indexOf(toReplace) === -1) {
+    throw new Error('wrapperify failed to find code to replace');
+  }
   src = src.replace(
-    `function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.${this.data.standalone} = f()}}`,
+    toReplace,
     `function(f){return f()}`
   );
   return `

